### ISSUE TITLE
Having nan in retinanet loss

### ISSUE THREAD

**Issue by limhj23** [keras-team/keras-io#586]
Hi,

I made my own training script with reference to a lot of your code for retinanet example. I appreciate your effort on it. 

While the data encoding part seems okay, I added an empty anchor box generator for images where there is no object to detect. It may sound weird but my task requires inputs to include images without any object to detect.
So here is what I did

```{python}
def encode_sample(self, image_shape, gt_boxes, cls_ids):
      """Creates box and classification targets for a single sample"""
      anchor_boxes = self._anchor_box.get_anchors(image_shape[1], image_shape[2])
      cls_ids = tf.cast(cls_ids, dtype=tf.float32)
      
      def _encode_object():
          matched_gt_idx, positive_mask, ignore_mask = self._match_anchor_boxes(
              anchor_boxes, gt_boxes
          )
          matched_gt_boxes = tf.gather(gt_boxes, matched_gt_idx)
          box_target = self._compute_box_target(anchor_boxes, matched_gt_boxes)
          matched_gt_cls_ids = tf.gather(cls_ids, matched_gt_idx)
          cls_target = tf.where(
              tf.not_equal(positive_mask, 1.0), -1.0, matched_gt_cls_ids
          )
          cls_target = tf.where(tf.equal(ignore_mask, 1.0), -2.0, cls_target)
          cls_target = tf.expand_dims(cls_target, axis=-1)
          return tf.concat([box_target, cls_target], axis=-1)

      def _encode_empty():
          num_anchors = tf.shape(anchor_boxes)[0]
          box_target = tf.zeros([num_anchors, 4], dtype = tf.float32)
          cls_target = tf.fill([num_anchors, 1], -1.0)
          return tf.concat([box_target, cls_target], axis=-1)

      label = tf.cond(
          tf.equal(tf.shape(input=cls_ids)[0], 0), 
          true_fn=lambda: _encode_empty(), 
          false_fn=lambda: _encode_object()
      )
      return label
```

It seems working okay with making the TF Dataset, but it seems not okay when it comes to the actual training. It produces NaN for loss and that happens for both cls_loss and box_loss.

I am not sure that this issue is caused by the empty anchor encoding, but just my guess. 
NaN values would probably mean gradient exploding, but the specific causes remain unknown. 

Please advise me

Regards
Ted

**Comment by csimmat** [comment]
Hi Ted,
I'm facing the same problem. Did you find something that could help?
Thanks,

Regards
Clara

**Comment by limhj23** [comment]
I used so many methods to try fixing it, but I think the main reasons were two below.

1. Gradient exploding: I used clip_norm option in my optimizer to clip the gradient to 1.0
2. Batch Label: I found some of the batches only contained background examples. Since focal loss doesn't count background in its calculation, you could choose either two
  a. Forcing every batch to have both background and foreground (object) images
  b. Change the loss function take account for background examples.

Maybe I am leaving this comment too late, but hope this helps.

**Comment by csimmat** [comment]
Hi Ted,
Thanks a lot for your answer. As you mentioned, it was due to batches of background examples only!
Good luck in your work
Clara

**Comment by SuryanarayanaY** [comment]
Hi @limhj23 ,

If it resolved for you can we mark it as closed? Please feel free to close the issue if resolved.
